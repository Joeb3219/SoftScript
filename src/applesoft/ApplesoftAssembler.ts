import { OpcodeToApplesoftInstructionMap } from "./Applesoft.types";
import fs from "fs";

export class ApplesoftAssembler {
  private static readonly _MAX_LINE_NUMBER: number = 65_536;
  private static readonly _STARTING_ADDRESS: number = 2048;

  constructor(private readonly lines: string[]) {}

  private assembleLine(currentAddress: number, line: string): number[] {
    const parts = line.split(" ");
    const [lineNumberStr, operation, ...dataStr] = parts;
    const lineNumber = lineNumberStr ? parseInt(lineNumberStr) : -1;

    if (lineNumber < 0 || lineNumber >= ApplesoftAssembler._MAX_LINE_NUMBER) {
      throw new Error(`Line number is invalid: ${lineNumber}`);
    }

    const operationOpcode = Object.entries(
      OpcodeToApplesoftInstructionMap
    ).find((e) => e[1] === operation)?.[0];

    if (!operationOpcode) {
      throw new Error(`Operation is invalid: ${operation}`);
    }

    const data = dataStr
      .join(" ")
      .split("")
      .map((c) => c.charCodeAt(0));

    // 2 bytes for the next instruction address,
    // 2 bytes for the line number
    // 1 byte for the operation,
    // data.length bytes for the data,
    // and 1 byte for the 0x00 byte to end instruction.
    const numBytes = 2 + 2 + 1 + data.length + 1;
    const buffer = Buffer.alloc(numBytes);

    const nextInstructionAddress = currentAddress + numBytes + 1;

    // Write all of the data byte by byte, just like it was meant to be.
    buffer.writeUint16LE(nextInstructionAddress, 0);
    buffer.writeUint16LE(lineNumber, 2);
    buffer.writeUint8(parseInt(operationOpcode), 4);
    for (let i = 0; i < data.length; i++) {
      buffer.writeUint8(data[i], 5 + i);
    }
    buffer.writeUint8(0x00, numBytes - 1);

    return [...buffer];
  }

  assemble(): number[] {
    const lineBytes = this.lines.reduce<number[]>((state, line) => {
      const currentAddress =
        state.length + ApplesoftAssembler._STARTING_ADDRESS;
      const newBytes = this.assembleLine(currentAddress, line);

      return [...state, ...newBytes];
    }, []);

    // The last two bytes of the program are always 0x00.
    return [...lineBytes, 0x00, 0x00];
  }
}
