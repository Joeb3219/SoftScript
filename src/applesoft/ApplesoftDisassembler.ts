import { OpcodeToApplesoftInstructionMap } from "./Applesoft.types";

type ApplesoftLine = {
    nextInstructionAddress: number;
    line: number;
    token: string;
    data?: any;
}

export class ApplesoftDisassembler {
    constructor(private readonly bytes: number[]) {}

    private disassembleLine(bytes: number[]): ApplesoftLine {
        return {
            nextInstructionAddress: (bytes[1] << 8) | (bytes[0]),
            line: (bytes[3] << 8) | (bytes[2]),
            token: OpcodeToApplesoftInstructionMap[bytes[4] as keyof typeof OpcodeToApplesoftInstructionMap],
            data: bytes.slice(5, bytes.length - 2).map(c => String.fromCharCode(c)).join('')
        }
    }

    private readInt16(idx: number) {
        return (this.bytes[idx + 1] << 8) | (this.bytes[idx]);
    }

    disassemble() {
        let currentAddress: number = 2048;
        const disassembledLines: ApplesoftLine[] = [];
        while ((currentAddress - 2048) < this.bytes.length && currentAddress > 0) {
            const idx = currentAddress - 2048;
            // First, we look at where the next instruction is located.
            const address = this.readInt16(idx);
            const instructionLength = address - currentAddress;

            const relevantBytes = this.bytes.slice(idx, idx + instructionLength);            
            const disassembled = this.disassembleLine(relevantBytes);
            disassembledLines.push(disassembled);

            currentAddress = disassembled.nextInstructionAddress;
        }

        console.log(disassembledLines);

        return disassembledLines;
    }
}