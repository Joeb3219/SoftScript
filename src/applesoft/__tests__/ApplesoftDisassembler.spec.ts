import { ApplesoftAssembler } from "../ApplesoftAssembler";
import { ApplesoftDisassembler } from "../ApplesoftDisassembler";

describe("ApplesoftDisassembler", () => {
    // These tests all write to a snapshot file that is hand-verified.
    describe("disassemble", () => {
        it("should correctly handle REM instructions", () => {
            const testData = [`1 REM Eat "your" = vegetables`];
            const assembler = new ApplesoftAssembler(testData);
            const assembled = assembler.assemble();
            const disassembler = new ApplesoftDisassembler(assembled);
            const result = disassembler.disassemble();
            expect(result.map((r) => r.fullInstruction)).toEqual(testData);
            expect(result).toMatchSnapshot();
        });

        it("should correctly handle string assignment", () => {
            const testData = [`1 LET X$ = "some value"`];
            const assembler = new ApplesoftAssembler(testData);
            const assembled = assembler.assemble();
            const disassembler = new ApplesoftDisassembler(assembled);
            const result = disassembler.disassemble();
            expect(result.map((r) => r.fullInstruction)).toEqual(testData);
            expect(result).toMatchSnapshot();
        });

        it("should correctly handle multiple statements", () => {
            const testData = [
                `1 LET X$ = "some value"`,
                `2 PRINT X$`,
                `3 LET Y$ = X$ + "some other test"`,
                `4 PRINT Y$`,
                `5 GOTO 1`,
            ];
            const assembler = new ApplesoftAssembler(testData);
            const assembled = assembler.assemble();
            const disassembler = new ApplesoftDisassembler(assembled);
            const result = disassembler.disassemble();
            expect(result.map((r) => r.fullInstruction)).toEqual(testData);
            expect(result).toMatchSnapshot();
        });
    });
});
