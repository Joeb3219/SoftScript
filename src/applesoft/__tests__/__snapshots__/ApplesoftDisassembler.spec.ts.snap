// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplesoftDisassembler disassemble should correctly handle REM instructions 1`] = `
[
  {
    "dataBytes": [
      178,
      32,
      69,
      97,
      116,
      32,
      34,
      121,
      111,
      117,
      114,
      34,
      32,
      61,
      32,
      118,
      101,
      103,
      101,
      116,
      97,
      98,
      108,
      101,
      115,
    ],
    "dataString": "REM Eat "your" = vegetables",
    "fullInstruction": "1 REM Eat "your" = vegetables",
    "line": 1,
    "nextInstructionAddress": 2079,
  },
]
`;

exports[`ApplesoftDisassembler disassemble should correctly handle multiple statements 1`] = `
[
  {
    "dataBytes": [
      170,
      88,
      36,
      208,
      34,
      115,
      111,
      109,
      101,
      32,
      118,
      97,
      108,
      117,
      101,
      34,
    ],
    "dataString": "LET X$ = "some value"",
    "fullInstruction": "1 LET X$ = "some value"",
    "line": 1,
    "nextInstructionAddress": 2070,
  },
  {
    "dataBytes": [
      186,
      88,
      36,
    ],
    "dataString": "PRINT X$",
    "fullInstruction": "2 PRINT X$",
    "line": 2,
    "nextInstructionAddress": 2078,
  },
  {
    "dataBytes": [
      170,
      89,
      36,
      208,
      88,
      36,
      200,
      34,
      115,
      111,
      109,
      101,
      32,
      111,
      116,
      104,
      101,
      114,
      32,
      116,
      101,
      115,
      116,
      34,
    ],
    "dataString": "LET Y$ = X$ + "some other test"",
    "fullInstruction": "3 LET Y$ = X$ + "some other test"",
    "line": 3,
    "nextInstructionAddress": 2107,
  },
  {
    "dataBytes": [
      186,
      89,
      36,
    ],
    "dataString": "PRINT Y$",
    "fullInstruction": "4 PRINT Y$",
    "line": 4,
    "nextInstructionAddress": 2115,
  },
  {
    "dataBytes": [
      171,
      49,
    ],
    "dataString": "GOTO 1",
    "fullInstruction": "5 GOTO 1",
    "line": 5,
    "nextInstructionAddress": 2122,
  },
]
`;

exports[`ApplesoftDisassembler disassemble should correctly handle string assignment 1`] = `
[
  {
    "dataBytes": [
      170,
      88,
      36,
      208,
      34,
      115,
      111,
      109,
      101,
      32,
      118,
      97,
      108,
      117,
      101,
      34,
    ],
    "dataString": "LET X$ = "some value"",
    "fullInstruction": "1 LET X$ = "some value"",
    "line": 1,
    "nextInstructionAddress": 2070,
  },
]
`;
